### ISSUE TITLE
Spike: Inventory and prioritize all existing Harvesting related issues

### ISSUE THREAD

**Issue by mreekie** [IQSS/dataverse#8574]
This is in support of:
- an NIH grant ["The Harvard Dataverse repository: A generalist repository integrated with a Data Commons"](https://docs.google.com/document/d/1cFK8pdwMKIRZxNs8EZXL1cXmW0ho6Rx9ulCewJ1ZCFI/edit?usp=sharing), 
  - Aim 4: **Improve harvesting and packaging standards to share metadata and data across repositories**, 

The first step is to figure out what has already been done by the dataverse team and by the community towards this aim and what still remains to be done. 

For example:
- [8139 Can't harvest when Dublin core field language is set](https://github.com/IQSS/dataverse/issues/8139#)

And then to prioritize which issues are to be fixed.

### Def of done
As completely as is reasonably possible in a 2 week period (sprint):
- [ ] Search out previous related issues that are problems with the current implementation. Take an inventory.
- [ ] Search out previous work done within the dataverse community as well.
- [ ] prioritize which of the issues/PRs that should be moved forward.

We need to keep in mind that to harvest something from a particular source requires that that source be bug free. Identify which sources have which bugs so that bugs for a particular source can be targeted. for example: ICPSR as an example. Zenodo is another.

### More information:

There is a lot packaged into Aim 4
 1. Improved Harvesting via the  OAI-PMH standard
 1. Improved support for Bagit
 1. Improved support for Signposting

The scope for this issue is Harvesting via the  OAI-PMH standard

### Aim 4: 

**Improve harvesting and packaging standards to share metadata and data across repositories**

Our proposed project will significantly improve the widely-used Harvard Dataverse repository to better support NIH-funded research. 

**A critical measure of the [GREI program](https://www.iq.harvard.edu/news/dataverse-joins-nih-data-repository-initiative)’s success is to standardize the discoverability across generalist repositories**.

To help with this, **we propose to improve the existing harvesting functionality in the Dataverse software based on the Open Archives Initiative Protocol for Metadata Harvesting **(OAI-PMH) standard**, and coordinate with other repository packaging standards to share or move metadata and data. **

Dataverse already supports the Bags as defined by the Research Data Alliance (RDA) Research Data Repository Interoperability Working Group. Here we proposed to improve the support for **Bags**, test it for NIH-funded datasets, and explore and define the appropriate standard to use to move the metadata and data across generalist repositories. This will help with a sustainable and succession plan - if one repository cannot support anymore a specific dataset, it will allow to easily move the dataset to another repository without losing any information about the dataset. 

Additionally we propose to implement **Signposting** in the Dataverse software.  By adding additional http link headers throughout the application, we can more easily support automated metadata and data discovery in the repository, and allow for other applications and services to more accurately and completely represent the content in the Harvard Dataverse repository. 

### Related documents
- [Notes on Dataverse Deliverablas for NIH OTA](https://docs.google.com/document/d/1N9xgubVcHb2mQxnCrmusa0M7qVGiVvyvmODc4j4HvQ8/edit?usp=sharing)
- [NIH OTA Progress Notes](https://docs.google.com/document/d/1k0XLOqYGCbV1O4eqUtOz67Hk4kXqhv4dKg_juabNVw0/edit?usp=sharing)
- [NIH OTA](https://docs.google.com/document/d/1cFK8pdwMKIRZxNs8EZXL1cXmW0ho6Rx9ulCewJ1ZCFI/edit?usp=sharing)
- [Exposing and harvesting metadata using the OAI metadata harvesting protocol: A tutoria (2001)](https://arxiv.org/pdf/cs/0106057.pdf)
- [Getting Started with BagIt in 2018](https://patchbay.tech/blog/2018/03/14/getting-started-with-bagit-in-2018/)
- [NIH OTA](https://docs.google.com/document/d/1cFK8pdwMKIRZxNs8EZXL1cXmW0ho6Rx9ulCewJ1ZCFI/edit?usp=sharing)
- [bagit from Library of Congress video](https://www.youtube.com/watch?time_continue=11&v=l3p3ao_JSfo&feature=emb_logo)

**Comment by pdurbin** [comment]
The definition of done for this issue includes three items:

- Search out previous related issues that are problems with the current implementation. Take an inventory.
- Search out previous work done within the Dataverse community as well.
- Prioritize which of the issues/PRs that should be moved forward.

I'm going to provide lists below for the first two items (open issues and merged pull requests). Please note that PR #7053 is an open pull request that addresses a harvesting issue. I'm not sure which list it belongs in but it closes #7502 which is in the open issues list.

The third item is about prioritizing what to work on. I'm leaving this for the prioritization group (@lenwiz, @scolapasta, @TaniaSchlatter, @sbarbosadataverse, and @mreekie). At standup today Gustavo did say that developers like me are welcome to give opinions on what we think should be the priority. From a quick look I'd say we should work on #5840 and #8267 because they are both small documentation updates. Otherwise, it's hard for me to judge the priority. I'm always for fixing bugs and there are plenty of small bugs.

I did reach out to the community to encourage them to open new issues about harvesting if they know of any:

- https://groups.google.com/g/dataverse-community/c/bDaOH8a4Hmc/m/XfzYL-BlAQAJ
- https://matrix.to/#/!AmypvmJtUjBesRrnLM:matrix.org/$xr-LkLmDvTolW0O4FoWZivjegovdGttxu3pBbnGisxk 
- https://dataversecommunity.slack.com/archives/C5V66TV6Y/p1649865300678319 

Here are the lists.

## Search out previous related issues that are problems with the current implementation. Take an inventory.

Here's the oneliner I used to get the following list of open issues from the GitHub Search API:

`curl -H 'Accept: application/vnd.github.v3.text-match+json' 'https://api.github.com/search/issues?q=is:issue%20is:open%20label:%22Feature:%20Harvesting%22%20repo:IQSS/dataverse&per_page=100' | jq '.items[] | "- #\(.number)"' -r`

- #8578
- #8575
- #8574
- #8512
- #8306
- #8290
- #8289
- #8267
- #8139
- #8129
- #8103
- #8005
- #7940
- #7736
- #7727
- #7624
- #7546
- #7453
- #7330
- #7243
- #7127
- #7108
- #7074
- #7073
- #7052
- #7021
- #6772
- #6275
- #5960
- #5920
- #5889
- #5840
- #5392
- #5112
- #4859
- #4831
- #4176
- #4025
- #3797
- #3741
- #3699
- #3648
- #3621
- #3437
- #3322
- #3316
- #3315
- #3314
- #3298
- #3290
- #3276
- #2721
- #2297
- #1814
- #1592

## Search out previous work done within the Dataverse community as well.

Here's the oneliner I used to get the following list of merged pull requests from the GitHub Search API:

`curl -H 'Accept: application/vnd.github.v3.text-match+json' 'https://api.github.com/search/issues?q=is:pr%20is:merged%20label:%22Feature:%20Harvesting%22%20repo:IQSS/dataverse&per_page=100' | jq '.items[] | "- #\(.number)"' -r`

- #8197
- #8106
- #7984
- #7974
- #7868
- #7741
- #7690
- #7214
- #7062
- #7057
- #7024
- #7010
- #7004
- #7002
- #6988
- #6935
- #6686
- #6669
- #6538
- #6513
- #6376
- #6087
- #5926
- #5834
- #5255
- #5248
- #5247
- #4787
- #4691
- #4664
- #4432
- #4096
- #4018
- #3521
- #3458
- #3308
- #2938
- #2933

**Comment by jggautier** [comment]
There are a few issues in the Harvard Dataverse Repository GitHub repo about harvesting:

- https://github.com/IQSS/dataverse.harvard.edu/issues/63
- https://github.com/IQSS/dataverse.harvard.edu/issues/72
- https://github.com/IQSS/dataverse.harvard.edu/issues/92
- https://github.com/IQSS/dataverse.harvard.edu/issues/142

And some support emails saved in the RT system about harvesting issues in the Harvard repository:

- https://help.hmdc.harvard.edu/Ticket/Display.html?id=287243
- https://help.hmdc.harvard.edu/Ticket/Display.html?id=295542
- https://help.hmdc.harvard.edu/Ticket/Display.html?id=311383
- https://help.hmdc.harvard.edu/Ticket/Display.html?id=314258

**Comment by landreev** [comment]
*(the lists below are work in progress, I'm actively working on them!)*

I do believe that the third item under the "definition of done" - "prioritize" - was the actual important part of this spike. I also believe that most of that effort of prioritizing what's important can only be done within the dev. team. I can't think of how anyone outside of it could be more qualified to make these calls. So I'm going to make such an attempt. 
(Note that I'm interpreting the word "prioritizing" as assigning some order of importance to these issues and bugs, what makes sense to fix first and/or what's ready to be worked on vs. what needs more discussion; not as scheduling them for specific sprints, etc.!)

The single most important harvesting issue: (ok, maybe not the most important - but seriously, this should be the first step of any meaningful cleanup of our harvesting implementation; should be fairly easy to wrap up too)
- https://github.com/IQSS/dataverse/issues/8372

The following issues are important in that fixing them will make harvesting more reliable and robust overall (for example, in the current implementation a *single* missing metadata export that's supposed to be cached is going to break the entire harvesting run). All of the issues on the list below are defined clearly enough that they are ready to be worked on and fixed, without needing to conduct any extra research first. Some of them may be VERY OLD; but they look like something we should fix. 
- https://github.com/IQSS/dataverse/issues/8484
- https://github.com/IQSS/dataverse/issues/7940
- https://github.com/IQSS/dataverse/issues/8005
- https://github.com/IQSS/dataverse/issues/7736
- https://github.com/IQSS/dataverse/issues/3437
- https://github.com/IQSS/dataverse/issues/3322

the following 3 issues are basically the same thing - people requesting extra ISO language codes to be added as legitimate controlled vocab. values (this is just a matter of adding extra values to citation.tsv); these are NOT duplicates, different things are being requested to be added in the issues below:
- https://github.com/IQSS/dataverse/issues/8578
- https://github.com/IQSS/dataverse/issues/8139
- https://github.com/IQSS/dataverse/issues/8243

The following issues are about the DDI exporter producing XML that is not valid under the schema. I would consider creating a wider scope umbrella issue, something like "Make sure our DDI is valid against the schema (and maybe add a real time validation step to the export?)
The first one is ready to be worked on, I believe:
- https://github.com/IQSS/dataverse/issues/7127
These *may* need some discussion:
- https://github.com/IQSS/dataverse/issues/3648
- https://github.com/IQSS/dataverse/issues/7108

Similarly, the following issues are requests for changes in how we export DC; I believe these need to be reviewed/discussed, perhaps together? 
- https://github.com/IQSS/dataverse/issues/8129
- https://github.com/IQSS/dataverse/issues/7243

The following issues are proposed changes to the design of the harvesting framework and/or metadata exports. Meaning this is something we probably need to discuss as a team, before we decide that these are good ideas and proceed to implement them. But IMO they are (I opened all of them 😄): 

- https://github.com/IQSS/dataverse/issues/8629
- https://github.com/IQSS/dataverse/issues/8630
- https://github.com/IQSS/dataverse/issues/8631
- https://github.com/IQSS/dataverse/issues/5960 (maybe?)

There is of course this issue that was opened for figuring out what needs to be added specifically for the NIH/GREI grant: 
- https://github.com/IQSS/dataverse/issues/8575
It is obviously super important and should be prioritized too; but it's more of a design discussion and research, rather than something we already can start coding. 

The list above is by no means complete. If an issue is not listed, it does not necessarily mean that it's not important. But the ones that are listed above should be a good subset to start with.

**Comment by landreev** [comment]
The list(s) above should allow us to start working on cleaning up and improving harvesting. 
There are still issues labeled "Feature: Harvesting" that need to be triaged (there are some that may have been resolved since they were opened; there are some that may not need to be addressed in the first place - but then we should close them). For the next stage of this work however, once the above issues are addressed, it'll probably be more important to figure out what needs to be resolved/what features need to be added specifically for the NIH/GREI project (the spike #8575 on the list above).

**Comment by scolapasta** [comment]
I'm going to go ahead and close this issue as the spike work of identifying and prioritizing is done above. We've started adding a few of these issues onto the board (in Next Sprint) and we'll continue to use this as a reference to add more as that work gets completed. But no need to keep the actual spike open.

**Comment by pdurbin** [comment]
One more:

- #8632

**Comment by mreekie** [comment]
Grooming note:
 - The content here needs to be looked at in more detail.
 - The gist I've got from this readthru is that there is plenty to fix in harvesting and we need to not just blindly fix what is here. 
 - It seems like there are items that are adjacent to harvesting that we could argue or not that need to be fixed to make harvesting work well.

**Comment by mreekie** [comment]
sizing:
- This is a time bounded spike.
- In this issue leonid proposes to review all of the known harvesting issues and make sense of them.  
- Once he is done we will have an ordered list and a decision on where to stop for the purposes of the NIH grant.

That decision will be reflected in a list in
- https://github.com/IQSS/dataverse/issues/8680

**Comment by mreekie** [comment]
Priority Review with Stefano:
- Moved from  NIH Deliverables Backlog to Ordered Backlog

**Comment by mreekie** [comment]
Sizing:
- We have changed (as of now) how we size spikes.
- The idea on this one is to invest 3 days (33)

**Comment by mreekie** [comment]
Daily
- Phil raised that he's gotten reports that harvesting breaks after upgrades.  He's been asking folks to create issues and include specifics on what versions they are using.
- Leonid mentioned that in general that folks should recreate their harvesting after upgrades.

**Comment by pdurbin** [comment]
> Phil raised that he's gotten reports that harvesting breaks after upgrades. He's been asking folks to create issues and include specifics on what versions they are using.

Yeah, I started a thread on Slack about this: https://iqss.slack.com/archives/C010LA04BCG/p1674160762504939

**Comment by landreev** [comment]
Here's the reordered/revisited list as it stands now. (_split in parts; work in progress_)

**1. Progress update.**

Harvesting issues that have been addressed/closed since the spike was created:

- #8843
- #8640
- #8372 
- #8267 
- #8139 
- #8103 
- #7940 
- #6275 
- #5840 
- #5112 
- #3797 
- #3741
- #3621

(Some of the issues above, those with low numbers especially, were closed having been fixed as part of more recent efforts/PRs. For example, #3797 was resolved as part of a general overhaul in #8372)

All of the above are being taken off the remaining prioritized list.

**Comment by landreev** [comment]
**2. Already estimated and/or prioritized:**

("prioritized" in this context means have at least been reviewed and deemed important/necessary to be addressed soon, and have been assigned NIH grant labels; and specifically `pm.epic.nih_harvesting`)

- #8005 (not sized yet)
- #7736 (10)
- #3437 (not sized)
- #3322 (not sized)
- #8289 (33)
- #8578 (not sized) (metadata import - "language" attribute)
- #8243 (not sized) (metadata import - "language" attribute)
- #3648 (33) (metadata export - DDI)
 
**2a. The following issues have NOT been tagged `pm.epic.nih_harvesting` yet** 
but they deal with metadata issues (see labels), and have been proposed for being addressed together with the metadata issues from the list above; so these could be good candidates for being pulled in soon.

- #7127 (not sized; NOT tagged `pm.epic.nih_harvesting`) (metadata export - DDI)
- #7108 (not sized; NOT tagged `pm.epic.nih_harvesting`) (metadata export - DDI)
- #8129 (not sized; NOT tagged `pm.epic.nih_harvesting`) (metadata export - DDI)
- #7243 (not sized; NOT tagged `pm.epic.nih_harvesting`) (metadata import/export - DC)

**Comment by landreev** [comment]
**3. Finally, some choice candidates for the short-term queuing, in roughly prioritized order:**

I would petition to prioritize this one: 

- #8842
But it is really too large, so I'm further proposing to split it into child issues:
- #9327 (this one is the most pressing, IMO; it should address things like
- #9294 in the process 

This one is a good candidate: 
This will ensure that our OAI gateway is valid per the spec to the outside clients (should be a good thing under the grant too?). The actual details/problems as described in the issue may no longer be true (resolved in the major reorganization of #8372). So the issue should be used to re-run the validation and fix whatever issues may remain (or may have been introduced as of late).
- #8815
 

The following one is a feature request that came in from an external contributor *with a fix PR accompanying it*; it may have got stuck in review limbo, we owe it to them to process it quickly:
- #8778

Same as above; small-ish issue, with an accompanying PR from an external dev. From having taken a look, it's not as trivial as they think. But should not be difficult to resolve and we owe it to them, etc. 
- #8739

An interesting feature request, very specific and should be easy to address. Can definitely be useful to other instances. Not sure if explicitly "useful" from the point of view of the NIH grant though. 
- #7330

Change in specific harvesting behavior requested; lots of detailed discussion in the issue. Should be ready to be addressed:
- #8693

This would be a genuinely useful format to add harvesting support for: 
- #2721 

Not strictly harvesting-related - but if we were willing to handle it under this umbrella, I would give it high priority:
- #8630

May already be resolved, but if not, def. a good thing to address promptly: 
- #4831

This is a very recent problem report from a remote installation. Their problem can be translated into a feature request, asking for more generic OAI_DC records (ones without persistent identifiers) to be importable. Could be a useful thing? - But I'm on a fence somewhat, about how it should be prioritized. 

- #9318

**Comment by mreekie** [comment]
Sprint review.
- This issue is going to be closed.
- The issues can be organized into the backlog with the NIH deliverables.

**Comment by landreev** [comment]
Oh, this is the one I forgot to mention: 
I'm not sure what to do with this huge jumbo issue that was recently opened:
- #8703
It has 4 child issues; 2 of them somewhat small-ish, that I have addressed above; but the remaining 2 are both potentially huge on their own:
- #8631
- #5960
These would be interesting features; but it's not clear to me if we can afford to prioritize them soon, when we have more pressing/important things to address. I'm going to mention it, in case anyone else feels strongly about them.

**Comment by landreev** [comment]
... and maybe I should explicitly list the remaining "bulk" queue - these are all labeled with the "feature: Harvesting" label. Basically, by omitting them, I communicated that these should not be in the next wave of prioritization/scheduling. But I'll post that remaining list, w/ my comments I was adding in the process.

**4. Remaining bulk list; I believe these can/should wait to be addressed, but opinions welcome.**

- #7546 (a lot/too much going on in this one; but I believe most of it will be addressed by issues in the already prioritized queues above)
- #8632 (specific behavior with a specific archive; hard to justify high priority)
- #8512 (license standartization request)
- #5920 (metadata issue... that I frankly don't fully understand, so this one may be up for debate/need a second expert opinion)
- #5889 (same as above)
- #3699 (same as above)
- #8306 (user report; some export/harvest failure)
- #7624 (cleaning up the client-side harvesting code may address this)
- #7453 (distantly related to harvesting - harvested datasets showing in saved searches funny?)
- #7074 (cleaning up the client-side harvesting code may address this)
- #7052 (cleaning up the client-side harvesting code may address this)
- #7021 (only related to harvesting because of bad metadata in some legacy harvested datasets in prod.)
- #5392 (not sure if it's actionable; aside from continuing to make harvesting more reliable overall, by addressing specific tech. issues)
- #4025 (simple linking feature - but not clear if there's still demand for it?)
- #3316 (very old harvesting server GUI issue; may or may not still be there; doesn't strike me as high priority)
- #3315 (old harvesting server-side issue; low-ish priority)
- #3314 (very old harvesting server GUI issue; may be nice to address eventually, but not super crucial)
- #3298 (I would just close this one - ?)
- #3290 (an ancient issue about displaying harvested search results - may or may not have been fixed; long discussion in the comments)
- #3276 (I don't think this is a Harvesting issue; and I don't think it's in the right queue - this would be better off in the prod. queue. But then it's from 2016, the results mentioned there are almost certainly not true anymore. So, idk, I would close this one; and maybe revisit the status of our prod. metadata exports from scratch.)
- #1814 (ancient issue - I'd be curious to find out if any of it is still relevant, if there's still interest in harvesting and/or importing NARA data?)
- #1592 (ancient issue - but I can see how it could be useful. Phil recently added a comment suggesting to open a new issue for adding support for DCAT, and close this one. DCAT is an alternative harvesting mechanism - alternative to OAI)

And just because I think this could be wait, for the purposes of the immediate planning, doesn't mean that I don't want them close. We should continue looking at this list. As I mention in some of the comments above, some of these could be fixed/resolved by the open issues we've already prioritized, so we'll need to confirm that and close those accordingly.
